src/org/intellij/erlang/psi/impl/ErlangPsiImplUtil.java

```diff
@@ -1839,38 +1882,91 @@ public ErlangFunctionCallParameter(@NotNull String funName, @NotNull String modu
-          ResolveResult[] results = reference instanceof PsiPolyVariantReference ?
-            ((PsiPolyVariantReference) reference).multiResolve(true) : ResolveResult.EMPTY_ARRAY;
-          for (ResolveResult r : results) {
-            if (expectedFunction(r.getElement())) return true;
-          }
-        }
+      ErlangFunctionCallExpression call = PsiTreeUtil.getParentOfType(element, ErlangFunctionCallExpression.class);
+      if (call == null || getPositionInFunctionCall(call, element) != myPosition) return false;
+      for (ErlangFunction function : resolveToFunctions(call)) {
+        if (expectedFunction(function)) return true;
       }
       return false;
     }
 
-    private boolean expectedFunction(@Nullable PsiElement resolve) {
-      if (!(resolve instanceof ErlangFunction)) return false;
-      String name = ((ErlangFunction) resolve).getName();
-      if (!name.equals(myFunName)) return false;
-      PsiFile containingFile = resolve.getContainingFile();
-      if (!(containingFile instanceof ErlangFile)) return false;
-      ErlangModule module = ((ErlangFile) containingFile).getModule();
+    private boolean expectedFunction(@NotNull ErlangFunction resolvedFunction) {
+      ErlangFile containingFile = ObjectUtils.tryCast(resolvedFunction.getContainingFile(), ErlangFile.class);
+      ErlangModule module = containingFile != null ? containingFile.getModule() : null;
       String moduleName = module != null ? module.getName() : null;
-      if (!myModule.equals(moduleName)) return false;
-      int arity = getArity((ErlangFunction) resolve);
-      if (arity != myArity) return false;
-      return true;
+      return myModule.equals(moduleName) && myFunName.equals(resolvedFunction.getName()) &&
+             myArity == getArity(resolvedFunction);
+    }
+  }
+
+  public static class ErlangFunctionCallModuleParameter extends PatternCondition<ErlangQAtom> {
+    public ErlangFunctionCallModuleParameter() {
+      super("functionCallModuleParameter");
+    }
+
+    @Override
+    public boolean accepts(@NotNull ErlangQAtom element, ProcessingContext context) {
+      ErlangFunctionCallExpression call = PsiTreeUtil.getParentOfType(element, ErlangFunctionCallExpression.class);
+      int position = call != null ? getPositionInFunctionCall(call, element) : -1;
+      if (position == -1) return false;
+
+      ErlangSpecification specification = findSpecification(resolveToFunction(call));
+      ErlangFunTypeSigs signature = specification != null ? specification.getSignature() : null;
+      if (signature == null) return false;
+
+      for (ErlangTypeSig typeSig : signature.getTypeSigList()) {
+        List<ErlangType> types = typeSig.getFunType().getFunTypeArguments().getTypeList();
+        if (types.size() <= position) continue;
+        ErlangType type = types.get(position);
+        ErlangQVar typeVar = type.getQVar();
+        if (argumentIsModule(typeVar != null ? getType(typeVar, typeSig.getTypeSigGuard()) : type)) {
```

Comment:
Introduce a variable for the actual type, or reassign an existing one - it will improve readability, and will aid during debugging this code in future.